[ML] Introduce generic dense embedding results classes

DonalEvans opened the pull request:
- Create new GenericDenseEmbedding\*Results classes which are identical to the existing DenseEmbedding\*Results classes, but which use "embeddings" instead of "text_embedding" in the results JSON
- Extract common code to abstract classes
- Duplicate tests for the new DenseEmbedding*Results classes
- Move the definition of the "embedding" constant String used by classes that implement the EmbeddingResult interface into that class
- Replace hard-coded uses of "text_embedding", "text_embedding_bytes", "text_embedding_bits", "sparse_embedding" and "embedding" fields with references to constants

DonalEvans committed:
Introduce generic dense embedding results classes

- Create new GenericDenseEmbedding*Results classes which are identical
  to the existing DenseEmbedding*Results classes, but which use
  "embeddings" instead of "text_embedding" in the results JSON
- Extract common code to abstract classes
- Duplicate tests for the new DenseEmbedding*Results classes
- Move the definition of the "embedding" constant String used by classes
  that implement the EmbeddingResult interface into that class
- Replace hard-coded uses of "text_embedding", "text_embedding_bytes",
  "text_embedding_bits", "sparse_embedding" and "embedding" fields with
  references to constants

DonalEvans committed:
Merge branch 'main' into create-multimodal-embeddings-results-classes

jonathan-buttner commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/AbstractDenseEmbeddingBitResults.java:
```diff
@@ -0,0 +1,92 @@
+        this(in.readCollectionAsList(AbstractDenseEmbeddingByteResults.Embedding::new));
+    }
+
+    public abstract String getArrayName();
```
What do you think about storing the array field name as a member of the abstract class and the child classes can define it by passing a static string via a constructor (like `super(TEXT_EMBEDDING_BITS)`). Then we can make this method `getArrayName` a member of the abstract class (or we might not need it, I'm not sure if it's called externally).

DonalEvans replied:
Yeah, I think this is better, since the array name doesn't need to be exposed outside of the abstract class.

jonathan-buttner reviewed (commented):
Looks good, left some questions

DonalEvans commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/GenericDenseEmbeddingBitResults.java:
```diff
@@ -0,0 +1,55 @@
+ */
+public final class GenericDenseEmbeddingBitResults extends AbstractDenseEmbeddingBitResults {
+    public static final String NAME = "embedding_bit_results";
+    public static final String EMBEDDINGS_BITS = "embeddings_bits";
```
For consistency between the three types of results, float, bytes and bits, the array names are `embeddings`, `embeddings_bytes` and `embeddings_bits`. Using `embedding` as the base of the array name would potentially cause problems since we already define each dense vector within the array as "embedding".

DonalEvans committed:
Address review comments

- Pass array name into abstract class constructor instead of overriding
  getter method
- Rename abstract classes to remove "AbstractDense" from the start
- Do not use TaskType to define array names for InferenceResults classes

DonalEvans committed:
Merge branch 'main' into create-multimodal-embeddings-results-classes
